lib/authz/permissions.go

```diff
@@ -332,6 +350,82 @@ func (a *authorizer) fromUser(ctx context.Context, userI interface{}) (*Context,
 	}
 }
 
+// checkAdminActionVerification checks if this auth request is verified for admin actions.
+func (a *authorizer) checkAdminActionVerification(ctx context.Context, authContext *Context) error {
+	if err := a.authorizeAdminAction(ctx, authContext); err != nil {
+		if trace.IsNotFound(err) {
+			// missing MFA verification should be a noop.
+			return nil
+		}
+		return trace.Wrap(err)
+	}
+
+	authContext.adminActionAuthorized = true
+	return nil
+}
+
+func (a *authorizer) authorizeAdminAction(ctx context.Context, authContext *Context) error {
+	// Builtin roles do not require MFA to perform admin actions.
+	switch authContext.Identity.(type) {
+	case BuiltinRole, RemoteBuiltinRole:
+		return nil
+	}
+
+	authpref, err := a.accessPoint.GetAuthPreference(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Admin actions do not require MFA when MFA is not enabled.
+	if authpref.GetSecondFactor() == constants.SecondFactorOff {
+		return nil
+	}
+
+	// Skip MFA check if the user is a Bot.
+	if user, err := a.accessPoint.GetUser(ctx, authContext.Identity.GetIdentity().Username, false); err == nil && user.IsBot() {
+		a.logger.Debugf("Skipping admin action MFA check for bot identity: %v", authContext.Identity.GetIdentity())
+		return nil
+	}
+
+	// Skip MFA if the identity is being impersonated by the Bot or Admin built in role.
+	if impersonator := authContext.Identity.GetIdentity().Impersonator; impersonator != "" {
+		impersonatorUser, err := a.accessPoint.GetUser(ctx, impersonator, false)
+		if err == nil && impersonatorUser.IsBot() {
+			a.logger.Debugf("Skipping admin action MFA check for bot-impersonated identity: %v", authContext.Identity.GetIdentity())
+			return nil
+		}
+
+		// If we don't find a user matching the impersonator, it may be the admin role impersonating.
+		// Check that the impersonator matches a host service FQDN - <host-id>.<clustername>
+		if trace.IsNotFound(err) {
+			hostFQDNParts := strings.SplitN(impersonator, ".", 2)
+			if hostFQDNParts[1] == a.clusterName {
+				if _, err := uuid.Parse(hostFQDNParts[0]); err == nil {
+					a.logger.Debugf("Skipping admin action MFA check for admin-impersonated identity: %v", authContext.Identity.GetIdentity())
+					return nil
+				}
+			}
+		}
+	}
+
+	// Certain hardware-key based private key policies require MFA for each request.
+	if authContext.Identity.GetIdentity().PrivateKeyPolicy.MFAVerified() {
+		return nil
+	}
+
+	// MFA is required to be passed through the request context.
+	mfaResp, err := mfa.CredentialsFromContext(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if _, _, err := a.accessPoint.ValidateMFAAuthResponse(ctx, mfaResp, authContext.User.GetName(), false); err != nil {
```

Comment:
Are we sure that MFA resp can be reused ? 
For instance for OPT tokens we have https://github.com/gravitational/teleport/blob/5d89de602447227eaa1f62fdaecc6369f1c7b032/lib/auth/password.go#L221 


lets say that `adminActionAuthorized` is required to create a teleport resource and the roles.yaml file contain multiple entries that will be handled by a sperate api call: 
roles.yaml
```
kind: role
version: v5
metadata:
  name: role1
--- 
kind: role
version: v5
metadata:
  name: role2
```

```
tctl create -f roles.yaml
```

In the result MFAAuthenticateResponse will be by 2x UpsertRole API calls.

Reply:
I think that is worth checking in final implementation. Retyping MFA for each `UpsertRole` call is not the best UX.